stan/math/prim/constraint/cholesky_corr_constrain.hpp

```diff
@@ -43,9 +43,11 @@ cholesky_corr_constrain(const EigVec& y, int K) {
 }
 
 // FIXME to match above after debugged
-template <typename EigVec, require_eigen_vector_t<EigVec>* = nullptr>
+template <typename EigVec, typename Lp,
+          require_eigen_vector_t<EigVec>* = nullptr,
+          require_convertible_t<return_type_t<EigVec>, Lp>* = nullptr>
```

Comment:
So this will allow `lp` to be `var` and `return_type_t<EigVec>` to be `double` or `var`.

Do we ever want to increment the jacobian when y just has a scalar type of `double`?

```cpp
lp += 0.5 * log1m(sum_sqs);
```

What are the use types you want to pass into this? Is it just the `return_type_t<EigVec> = double` and `lp = var` case? Since the data is considered constant in that case shouldn't we just call the one that does not increment the jacobian?

Reply:
> `return_type_t<EigVec> = double` and `lp = var`

Yes, that's the relevant one that comes up during code generation. Based on my conversation with @bob-carpenter, I think we still want to let the user do it, even if it's weird. The implementation in stanc3#1494 gives the user both a `_jacobian` function and a `_constrain` function, and I think it's reasonable to view these as similar to the difference between doing a target+= with a `_lpdf`  versus a `_lupdf`.